### ISSUE TITLE
resolver: Insecure DNSSEC delegations with minor faults cause TXT records in zone to fail

### ISSUE THREAD

**Issue by johnhtodd** [PowerDNS/pdns#4735]
- Program: Recursor
 - Issue type: Bug report

### Short description
TXT records for DMARC are failing in a zone that is insecurely delegated to a DMARC vendor who uses a nameserver that replies with unexpected but probably sufficient results.  The primary domain above the _dmarc. split is DNSSEC-secured.

### Environment
 - Operating system: CentOS
 - Software version:  7.2.1511 (pdns recursor 4.0.3)
 - Software source: PowerDNS repository

### Steps to reproduce

Step 1:Perform this lookup against PDNS recursor :  dig @localhost _dmarc.otalliance.org TXT

### Expected behavior

I would expect to receive the result:

`_dmarc.otalliance.org.	598	IN	TXT	"v=DMARC1\; p=reject\; rua=mailto:admin@otalliance.org,mailto:dmarc_agg@auth.returnpath.net,mailto:dmarc_agg@vali.email,mailto:online-trust-alliance@rua.agari.com\; ruf=mailto:admin@otalliance.org,mailto:dmarc_afrf@auth.returnpath.net,mailto:dmarc_dcfd0654_af" "rf@vali.email,mailto:online-trust-alliance@ruf.agari.com"`

### Actual behaviour
I receive a "SERVFAIL" response.

### Other information
Tests against unbound resolver on the same machine return the expected record, with DNSSEC validation turned on.

The DNSVIZ page shows multiple problems with the records in the zone after insecure delegation to ns.vali.email, notably that DS and DNSKEY records in the _dmarc zone return with a "REFUSED" response instead of "NOERROR".  However, the TXT, SOA, and NS records are valid and seem to be enough to create a correct response, as the zone was insecurely delegated.  Broken-ness on the remote auth resolver I don't think is sufficient to result in a total failure on lookup.

Discussion on IRC (2016/12/4 ~15:30 -08:00) by Maik & jtodd.

**Comment by pieterlexis** [comment]
note:

```
Dec 28 12:56:32 got 4 keys and 1 sigs from server
Dec 28 12:56:32 got valid DNSKEY (it matches the DS) with tag 2860 for otalliance.org
Dec 28 12:56:32 got valid DNSKEY (it matches the DS) with tag 55597 for otalliance.org
Dec 28 12:56:32 validation succeeded - whole DNSKEY set is valid
Dec 28 12:56:32 situation: we have one or more valid DNSKEYs for [otalliance.org] (want [_dmarc.otalliance.org])
Dec 28 12:56:32 next name [_dmarc.otalliance.org], trying to get DS
Dec 28 12:56:32 [1] _dmarc.otalliance.org: Wants DNSSEC processing in query for DS
Dec 28 12:56:32 [1] _dmarc.otalliance.org: No cache hit for '_dmarc.otalliance.org|DS', trying to find an appropriate NS record
Dec 28 12:56:32 [1] otalliance.org: Checking if we have NS in cache for 'otalliance.org'
Dec 28 12:56:32 [1] otalliance.org: NS (with ip, or non-glue) in cache for 'otalliance.org' -> 'pdns06.domaincontrol.com'
Dec 28 12:56:32 [1] otalliance.org: within bailiwick: 0, not in cache / did not look at cache
Dec 28 12:56:32 [1] otalliance.org: NS (with ip, or non-glue) in cache for 'otalliance.org' -> 'pdns05.domaincontrol.com'
Dec 28 12:56:32 [1] otalliance.org: within bailiwick: 0, not in cache / did not look at cache
Dec 28 12:56:32 [1] otalliance.org: We have NS in cache for 'otalliance.org' (flawedNSSet=0)
Dec 28 12:56:32 [1] _dmarc.otalliance.org: Cache consultations done, have 2 NS to contact
Dec 28 12:56:32 [1] _dmarc.otalliance.org.: Nameservers: pdns05.domaincontrol.com.(21.07ms), pdns06.domaincontrol.com.(22.19ms)
Dec 28 12:56:32 [1] _dmarc.otalliance.org: Trying to resolve NS 'pdns05.domaincontrol.com' (1/2)
Dec 28 12:56:32 [1]    pdns05.domaincontrol.com: Wants DNSSEC processing in query for A
Dec 28 12:56:32 [1]    pdns05.domaincontrol.com: Found cache hit for A: 216.69.185.52[ttl=599] 
Dec 28 12:56:32 [1] _dmarc.otalliance.org: Resolved 'otalliance.org' NS pdns05.domaincontrol.com to: 216.69.185.52
Dec 28 12:56:32 [1] _dmarc.otalliance.org: Trying IP 216.69.185.52:53, asking '_dmarc.otalliance.org|DS'
Dec 28 12:56:32 [1] _dmarc.otalliance.org: Got 4 answers from pdns05.domaincontrol.com (216.69.185.52), rcode=0 (No Error), aa=0, in 28ms
Dec 28 12:56:32 [1] _dmarc.otalliance.org: accept answer '_dmarc.otalliance.org|NS|ns.vali.email.' from 'otalliance.org' nameservers? 2 YES!
Dec 28 12:56:32 [1] _dmarc.otalliance.org: accept answer 'mr93206tk36mdprafodjlpnublt0utf6.otalliance.org|NSEC3|1 0 1 - OPHK96UER39PL2OJVKF2IQ700RTKJD0I NS' from 'otalliance.org' nameservers? 2 YES!
Dec 28 12:56:32 [1] _dmarc.otalliance.org: accept answer 'mr93206tk36mdprafodjlpnublt0utf6.otalliance.org|RRSIG|NSEC3 8 3 600 20170101185958 20161217185958 32568 otalliance.org. S+HIucGZwH3YR6Gl1UJPgMIMeLFJQvGW7+Lu8uPuSoRjK1fNxKZIvHEABw3AepScyi+i04isgIT3MmIT0q3PJnl3X3HiAhp0385aweM1rGCtL6v3oFMSyU8WjMnvDaWqwwsMlDhvnmwwGedXfrsqHhTQtDsFKHK0KUgSjCCVR0A=' from 'otalliance.org' nameservers? 2 RRSIG - separate
```

When asking for the DS records, we get a delegation instead of an NSEC record, leading us to follow this (false) delegation. All these nameservers return REFUSED, leading us to believe that the DS records don't exist (they don't) but there are no NSEC and RRSIGs to prove this.

The otalliance.org nameservers should return a NOERROR response with an NSEC(3) record indication that the name `_dmarc.otalliance.org` exists but not the `DS` type.

I think unbound responds this way because it tries somewhat harder to get an answer. Where we prefer 'correct'... I'm not sure how this could (or should) be fixed.. other then " make your auths do the correct thing"

**Comment by Habbie** [comment]
@johnhtodd can you check if current master still has trouble with this?

**Comment by johnhtodd** [comment]
This record resolves correctly with 4.1.0-0.1.rc2.1pdns.el7 @powerdns-rec-41

**Comment by pieterlexis** [comment]
Wonderful!
